File: pgpainless-core/src/main/java/org/pgpainless/key/generation/type/xdh/XDH.java
// SPDX-FileCopyrightText: 2020 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation.type.xdh;

import java.security.spec.AlgorithmParameterSpec;

import org.bouncycastle.jce.spec.ECNamedCurveGenParameterSpec;
import org.pgpainless.algorithm.PublicKeyAlgorithm;
import org.pgpainless.key.generation.type.KeyType;

public final class XDH implements KeyType {

    private final XDHSpec spec;

    private XDH(XDHSpec spec) {
        this.spec = spec;
    }

    public static XDH fromSpec(XDHSpec spec) {
        return new XDH(spec);
    }

    @Override
    public String getName() {
        return "XDH";
    }

    @Override
    public PublicKeyAlgorithm getAlgorithm() {
        return PublicKeyAlgorithm.ECDH;
    }

    @Override
    public AlgorithmParameterSpec getAlgorithmSpec() {
        return new ECNamedCurveGenParameterSpec(spec.getName());
    }

}

File: pgpainless-core/src/main/kotlin/org/pgpainless/key/generation/type/xdh/XDH.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation.type.xdh

import org.bouncycastle.jce.spec.ECNamedCurveGenParameterSpec
import org.pgpainless.algorithm.PublicKeyAlgorithm
import org.pgpainless.key.generation.type.KeyType

/** Legacy ECDH with Curve25519. For use with OpenPGP v4 only. */
class XDH private constructor(spec: XDHSpec) : KeyType {
    override val name = "XDH"
    override val algorithm = PublicKeyAlgorithm.ECDH
    override val bitStrength = spec.bitStrength
    override val algorithmSpec = ECNamedCurveGenParameterSpec(spec.algorithmName)

    companion object {
        @JvmStatic fun fromSpec(spec: XDHSpec) = XDH(spec)
    }
}
